sqlitecluster/SQLiteNode.cpp

```diff
@@ -1665,51 +1716,57 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
-                    if (message.calcU64("NewCount") != _db.getCommitCount() + 1) {
-                        STHROW("commit count mismatch. Expected: " + message["NewCount"] + ", but would actually be: "
-                              + to_string(_db.getCommitCount() + 1));
-                    }
-                    if (peer->permaFollower) {
-                        STHROW("permafollowers shouldn't approve/deny");
+
+            // If it's DENY or AsyncNotification is set means that it's not just a simple notification that the follower has some commit number.
+            // It's either a real DENY, or a real APPROVE of a quorum transaction.
+            if (SIEquals(message.methodLine, "DENY_TRANSACTION") || !message.isSet("AsyncNotification")) {
+                if (!message.isSet("ID")) {
+                    STHROW("missing ID");
+                }
+                if (!message.isSet("NewCount")) {
+                    STHROW("missing NewCount");
+                }
+                if (!message.isSet("NewHash")) {
+                    STHROW("missing NewHash");
+                }
+                if (_state != SQLiteNodeState::LEADING && _state != SQLiteNodeState::STANDINGDOWN) {
+                    STHROW("not leading");
+                }
+                SQLitePeer::Response response = SIEquals(message.methodLine, "APPROVE_TRANSACTION") ? SQLitePeer::Response::APPROVE : SQLitePeer::Response::DENY;
+                try {
+                    // We ignore late approvals of commits that have already been finalized. They could have been committed
+                    // already, in which case `_lastSentTransactionID` will have incremented, or they could have been rolled
+                    // back due to a conflict, which would cause them to have the wrong hash (the hash of the previous attempt
+                    // at committing the transaction with this ID).
+                    bool hashMatch = message["NewHash"] == _db.getUncommittedHash();
+                    if (hashMatch && to_string(_lastSentTransactionID + 1) == message["ID"]) {
+                        if (message.calcU64("NewCount") != _db.getCommitCount() + 1) {
+                            STHROW("commit count mismatch. Expected: " + message["NewCount"] + ", but would actually be: "
+                                  + to_string(_db.getCommitCount() + 1));
+                        }
+                        if (peer->permaFollower) {
+                            STHROW("permafollowers shouldn't approve/deny");
+                        }
+                        PINFO("Peer " << response << " transaction #" << message["NewCount"] << " (" << message["NewHash"] << ")");
+                        peer->transactionResponse = response;
+                    } else {
+                        // Old command.  Nothing to do.  We already sent a commit or rollback.
+                        // Shut this up.
```

Comment:
?

Reply:
Oh, it's noisy. I should reduce the logging of the following line.